3rdParty/fuerte/src/H1Connection.cpp

```diff
@@ -222,34 +224,52 @@ size_t H1Connection<ST>::requestsLeft() const {
   return q;
 }
 
+template <SocketType ST>
+void H1Connection<ST>::activate() {
+  FUERTE_ASSERT(_active.load());
+  Connection::State state = this->_state.load();
+  // FUERTE_ASSERT(state != Connection::State::Connecting);
+  if (state == Connection::State::Connected) {
+    FUERTE_LOG_HTTPTRACE << "activate: connected\n";
+    this->startWriting();
+  } else if (state == Connection::State::Disconnected) {
+    FUERTE_LOG_HTTPTRACE << "activate: not connected\n";
+    this->startConnection();
+  } else if (state == Connection::State::Failed) {
+    FUERTE_LOG_ERROR << "activate: queued request on failed connection\n";
+    drainQueue(fuerte::Error::ConnectionClosed);
+    _active.store(false);    // No more activity from our side
+  }
+  // If the state is `Connecting`, we do not need to do anything, this can
+  // happen if this `activate` was posted when the connection was still
+  // `Disconnected`, but in the meantime the connect has started.
+}
+
 template <SocketType ST>
 void H1Connection<ST>::finishConnect() {
+  // Note that the connection timeout alarm has already been disarmed.
+  // If it has already gone off, we might have a completion handler
+  // already posted on the iocontext. However, this will not touch anything
+  // if we have first set the state to `Connected`.
+  FUERTE_ASSERT(_active.load());
   auto exp = Connection::State::Connecting;
   if (this->_state.compare_exchange_strong(exp, Connection::State::Connected)) {
-    startWriting();  // starts writing queue if non-empty
+    this->startWriting();  // starts writing queue if non-empty
+  } else {
+    FUERTE_LOG_ERROR << "finishConnect: found state other than 'Connecting'";
+    FUERTE_ASSERT(false);
+    // If this happens, we probably have a sleeping barber
   }
 }
 
-// Thread-Safe: activate the combined write-read loop
+// startWriting must only be executed on the IO thread and only with active==true
+// and with state `Connected`
 template <SocketType ST>
 void H1Connection<ST>::startWriting() {
```

Comment:
can this be completely inlined now ?

Reply:
I think it can, the compiler should take care of this, at least with LTO. I only left it in to keep the similarity of the method names the same with H2 and VST.